flax/testing/benchmark.py

```diff
@@ -0,0 +1,241 @@
+    self.report_metrics({name: value})
+
+  def report_extras(self, extras: Dict[str, str]):
+    """Report extras for the benchmark."""
+    self._update_reported_name()
+    self._reported_extras.update(extras)
+
+  def report_extra(self, name: str, value: str):
+    """Report a single extra for the benchmark."""
+    self.report_extras({name: value})
+
+  def _get_test_name(self, prefix='test_'):
+    """Returns full name of test class and method calling report_benchmark.
+
+    The name is based on the *outermost* Benchmark class in the class stack.
+    Based on tensorflow/python/platform/benchmark.py
+
+    Args:
+      prefix: str. Prefix that the caller method must have.
+
+    Returns:
+      Resolved test name as `ClassName.test_name`.
+    """
+    # Find the caller method (outermost Benchmark class).
+    stack = inspect.stack()
+    calling_class, name = None, None
+    for frame_info in stack[::-1]:
+      # print(frame_info.frame, frame_info.function)
+      f_locals = frame_info.frame.f_locals
+      f_self = f_locals.get('self', None)
+      if isinstance(f_self, Benchmark):
+        name = frame_info.function
+        if name.startswith(prefix):
+          calling_class = f_self
+          break
+    if calling_class is None:
+      raise ValueError('Unable to determine the calling Benchmark class.')
+
+    # Prefix the name with the class name.
+    class_name = type(calling_class).__name__
+    name = "%s.%s" % (class_name, name)
+    return name
+
+  def _update_reported_name(self):
+    """Record / update test name for the benchmark."""
+    self._reported_name = self._reported_name or self._get_test_name()
+
+  def _report_benchmark_results(self):
+    """Produce benchmark results report.
+
+    Results are reported as a JSON string with the following schema:
+    ```
+    {
+      "name": <class.testMethod>
+      "succeeded": true / false
+      "wall_time": float (containing wall-time for the benchmark)
+      "metrics": {
+        "string" -> float map of other performance metrics
+      }
+      "extras": {
+        "string" -> "string" map containing anything else of interest
+      }
+    }
+    ```
+    """
+    name = self._reported_name
+    if not name:
+      raise ValueError('Unable to determine test name for reporting '
+                       'benchmark results. Make sure you are using '
+                       '`self.report_*` methods.')
+
+    succeeded = not self.has_outstanding_fails()
+    results = {'name': name,
+               'succeeded': succeeded,
+               'metrics': self._reported_metrics,
+               'extras': self._reported_extras}
+    if self._reported_wall_time is not None:
+      results['wall_time'] = self._reported_wall_time
+    if not succeeded:
+      msg = '\n'.join([str(fail) for fail in self._outstanding_fails])
+      results['extras']['failed_assertions'] = msg
+
+    results_str = json.dumps(results)
+    logging.info(results_str)
+    print(results_str)
```

Comment:
Is this left intentionally?

Reply:
Semi-intentionally - I could not figure out how to get pytest to print info-level logs. Removed the print statement now.